crates/rome_js_analyze/src/analyzers/correctness/no_unreachable_super.rs

```diff
@@ -107,140 +112,135 @@ impl Rule for NoUnreachableSuper {
-            let visited = predecessors
-                .entry(block_id)
-                .get_or_insert_with(BlockMap::<&JsSuperExpression>::default);
-
-            let previous_node = visited
-                .insert(prev_block, super_expression)
-                .filter(|previous_node| *previous_node == super_expression);
-
-            if previous_node.is_some() {
+        // We traverse the control flow graph and register in a context if we saw `super`.
+        // The context follows the traversal of every path (we process [BlockContext]).
+        //
+        // A block may be visiter twice: one time with an empty context,
+        // and another time with a `super` in its context.
+        //
+        // As soon as we detect the use of this without calling first `super`
+
+        // stack of block contexts to process
+        let mut block_context_stack = Vec::new();
+        let mut visited_block_contexts = FxHashSet::default();
+        block_context_stack.push(BlockContext {
+            block_index: 0,
+            super_: None,
+        });
+        // Set of couples (block_index, was_visited_with_super_call)
+        visited_block_contexts.insert((0u32, false));
+        while let Some(BlockContext {
+            block_index,
+            mut super_,
+        }) = block_context_stack.pop()
+        {
+            let had_super = super_.is_some();
+            // SAFETY: this is a safe conversion because it is already an index for `cfg.blocks`.
+            let block_index = block_index as usize;
+            let Some(block) = cfg.blocks.get(block_index) else {
                 continue;
-            }
-
-            if let Some(outgoing_edges) = outgoing_edges.get(block_id) {
-                for next_block in outgoing_edges {
-                    queue.push_back((next_block, block_id, super_expression));
-                }
-            }
-        }
-
-        // Check all the blocks containing a `super()` expression and emit an
-        // error if they have a predecessor (as it means `super()` may have
-        // already been called)
-        for (block_id, second) in &super_blocks {
-            if let Some(predecessors) = predecessors.get(block_id) {
-                if let Some(first) = predecessors.values().next() {
-                    return Some(RuleState::DuplicateSuper {
-                        first: first.syntax().text_trimmed_range(),
-                        second: second.syntax().text_trimmed_range(),
-                    });
+            };
+            for instruction in block.instructions.iter() {
+                if let Some(NodeOrToken::Node(ref block_node)) = instruction.node {
+                    let mut iter = block_node.preorder();
+                    while let Some(event) = iter.next() {
+                        let WalkEvent::Enter(node) = event else {
+                            continue;
+                        };
```

Comment:
I still need to get used to `let/else` 😆

Reply:
The structure is a bit strange because the else branch cannot be an expression.